### PR TITLE
feat: Add theme options for cheatsheet

### PR DESCRIPTION
**Pull request by Revathyvenugopal162**

### CODE REVIEW

File: doc/source/user_guide/options.rst [hunk 1]
```diff
@@ -188,3 +188,55 @@ to provide search functionality for your documentation.
...
+To enable the cheatsheet, you can add a child 
+dictionary called ``cheatsheet`` to the ``html_theme_options``
+dictionary in the ``conf.py`` file. 
+
+The dict should contain the following keys:
+
+#. ``url``: The URL of the cheatsheet for downloading.
+#. ``title``: Title of the cheatsheet.
+#. ``image``: The thumbnail image for the cheatsheet.
+#. ``local_download``: Specifies whether to download the cheatsheet locally during the build.
```
Comment (jorgepiloto) [inline doc/source/user_guide/options.rst:207]: Can I suggest to use a name more compliant with boolean variables? Something like:

```suggestion
#. ``needs_download``: Specifies whether to download the cheatsheet locally during the build. 
```

File: doc/source/user_guide/options.rst [hunk 2]
```diff
@@ -188,3 +188,55 @@ to provide search functionality for your documentation.
...
+
+To enable the cheatsheet, you can add a child 
+dictionary called ``cheatsheet`` to the ``html_theme_options``
+dictionary in the ``conf.py`` file. 
+
+The dict should contain the following keys:
+
+#. ``url``: The URL of the cheatsheet for downloading.
+#. ``title``: Title of the cheatsheet.
+#. ``image``: The thumbnail image for the cheatsheet.
```
Comment (jorgepiloto) [inline doc/source/user_guide/options.rst:206]: I think thumbnail describes better the purpose of this variable:

```suggestion
#. ``thumbnail``: The thumbnail image for the cheatsheet.
```

**Review by jorgepiloto — CHANGES_REQUESTED** [review]
Awesome impelementation! I just left some minor requests. Maybe some other team members want to also add new suggestions.

File: src/ansys_sphinx_theme/__init__.py [hunk 3]
```diff
@@ -285,6 +286,80 @@ def update_footer_theme(
...
+
+    Parameters
+    ----------
+    url : str
+        The URL of the file to download.
+    directory : str
+        The directory where the file will be saved.
+    """
+    filename = url.split("/")[-1]
+    if not filename in os.listdir(directory):
```
[the remark this answers is not in the document]
Comment (Revathyvenugopal162) [reply src/ansys_sphinx_theme/__init__.py]: adapted in https://github.com/ansys/ansys-sphinx-theme/pull/346/commits/5acf27d18e5301c1991d04e97606642493404996 and the commits after.

File: src/ansys_sphinx_theme/__init__.py [hunk 4]
```diff
@@ -285,6 +286,80 @@ def update_footer_theme(
...
+        file_path = directory / filename
```
[the remark this answers is not in the document]
Comment (Revathyvenugopal162) [reply src/ansys_sphinx_theme/__init__.py:364]: adapted in https://github.com/ansys/ansys-sphinx-theme/pull/346/commits/5acf27d18e5301c1991d04e97606642493404996 .

File: src/ansys_sphinx_theme/theme/ansys_sphinx_theme/components/cheatsheet_sidebar.html [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+theme_cheatsheet.get('title', 'Cheatsheet') %}
+<div class="sidebar-cheatsheets">
+  <h4>{{ theme_cheatsheet_title }}</h4>
+  {% if theme_cheatsheet.get('local_download')==true %} {% set
+  theme_cheatsheet_url = theme_cheatsheet.get('url').split('/')[-1] %} {% set
+  theme_cheatsheet_image = theme_cheatsheet.get('image').split('/')[-1] %} {%
+  set theme_cheatsheet_image = pathto('_static/' ~ theme_cheatsheet_image, 1) %}
+  {% set theme_cheatsheet_url = pathto('_static/' ~ theme_cheatsheet_url, 1) %}
+  {% else %} {% set theme_cheatsheet_url = theme_cheatsheet.get('url') %} {% set
+  theme_cheatsheet_image = theme_cheatsheet.get('image') %} {% endif %}
```
[the remark this answers is not in the document]
Comment (Revathyvenugopal162) [reply src/ansys_sphinx_theme/theme/ansys_sphinx_theme/components/cheatsheet_sidebar.html]: adapted in https://github.com/ansys/ansys-sphinx-theme/pull/346/commits/5acf27d18e5301c1991d04e97606642493404996.

File: doc/source/conf.py [hunk 6]
```diff
@@ -65,6 +65,12 @@
             f"ansys-sphinx-theme-v{convert_version_to_pymeilisearch(__version__)}": "ansys-sphinx-theme",  # noqa: E501
         },
     },
+    "cheatsheet": {
+        "url": "https://cheatsheets.docs.pyansys.com/pyaedt_API_cheat_sheet.pdf",
+        "title": "PyAEDT Cheatsheet",
+        "thumbnail": "https://cheatsheets.docs.pyansys.com/pyaedt_API_cheat_sheet.png",
+        "needs_download": True,
+    },
```
Comment (Revathyvenugopal162) [inline doc/source/conf.py:73]:
```suggestion
```

File: doc/source/user_guide/options.rst [hunk 7]
```diff
@@ -188,3 +188,55 @@ to provide search functionality for your documentation.
...
+The dict should contain the following keys:
+
+#. ``url``: The URL of the cheatsheet for downloading.
+#. ``title``: Title of the cheatsheet.
+#. ``thumbnail``: The thumbnail image for the cheatsheet.
+#. ``needs_download``: Specifies whether to download the cheatsheet locally during the build. 
+   If set to `True`, the cheatsheet is downloaded to `_build/html/_static/`, otherwise, 
+   it is accessed directly from the provided URL. ``needs_download`` defaults to `False`.
+#. ``pages``: A list of pages to include in the cheatsheet (optional). If not provided, 
+   the cheatsheet includes the index page of the documentation.
```
Comment (PipKat) [inline doc/source/user_guide/options.rst:211]:
```suggestion
Cheat sheets
------------

The ``cheatsheet`` HTML theme option can be used to enable a cheat sheet in the Ansys Sphinx theme.

To enable a cheatsheet, in the ``conf.py`` file for your PyAnsys
library, you can add a child dictionary named ``cheatsheet`` to
the ``html_theme_options`` dictionary. 

This dictionary should contain these keys, in the order given:

#. ``url``: URL of the cheat sheet for downloading.
#. ``title``: Title of the cheat sheet.
#. ``thumbnail``: Thumbnail image for the cheat sheet.
#. ``needs_download``: Whether to download the cheat sheet locally during the build. The default is ``False``, in which case the cheat sheet is accessed directly from the provided URL. If ``True``, the cheat sheet is downloaded to the ``_build/html/_static/`` directory. 
#. ``pages``: List of pages to include in the cheat sheet (optional). If no list is provided, the cheat sheet includes the index page of the documentation.
```
Cheat sheet (2 words) in descriptive text.
In second-to-last-bullet, does "build" refer to the documentation build? Maybe this should be more specific. In the last bullet, I'm not understanding what you mean. What list of pages can be included in the cheat sheet? How does the cheat sheet include the index page of the documentation. Do you mean "List of pages to include the cheat sheet on."?  I think that this is likely.

File: doc/source/user_guide/options.rst [hunk 8]
```diff
@@ -188,3 +188,55 @@ to provide search functionality for your documentation.
...
+.. code-block:: python
+
+    html_theme_options = (
+        {
+            "cheatsheet": {
+                "url": "<your cheatsheet URL>",
+                "title": "<title of your cheatsheet>",
+                "thumbnail": "<image URL>",
+                "needs_download": True,  # True if you want to download the cheatsheet locally in ``_build/html/_static/``
+                "pages": "<list of pages to include in the cheatsheet>",  # Optional
```
Comment (PipKat) [inline doc/source/user_guide/options.rst:222]:
```suggestion
                "url": "<your cheat sheet URL>",
                "title": "<title of your cheat sheet>",
                "thumbnail": "<image URL>",
                "needs_download": True,  # True if you want to download the cheatsheet locally in the ``_build/html/_static/`` directory.
                "pages": "<list of pages to include the cheat sheet on>",  # Optional
```

File: src/ansys_sphinx_theme/__init__.py [hunk 9]
```diff
@@ -285,6 +286,92 @@ def update_footer_theme(
     context["ansys_sphinx_theme_version"] = __version__
 
 
+def add_cheat_sheet(
+    app: Sphinx, pagename: str, templatename: str, context: Dict[str, Any], doctree: document
+) -> None:
+    """Add a cheatsheet to the sidebar.
```
Comment (PipKat) [inline src/ansys_sphinx_theme/__init__.py:292]:
```suggestion
    """Add a cheat sheet to the left navigation sidebar.
```

File: src/ansys_sphinx_theme/__init__.py [hunk 10]
```diff
@@ -285,6 +286,92 @@ def update_footer_theme(
...
+    app : ~sphinx.application.Sphinx
+        Application instance for rendering the documentation.
+    pagename : str
+        The name of the current page.
+    templatename : str
+        The name of the template being used.
+    context : dict
+        The context dictionary for the page.
+    doctree : ~docutils.nodes.document
+        The doctree
```
Comment (PipKat) [inline src/ansys_sphinx_theme/__init__.py:305]:
```suggestion
        Name of the current page.
    templatename : str
        Name of the template being used.
    context : dict
        Context dictionary for the page.
    doctree : ~docutils.nodes.document
        The doctree.
```
The description for the last bullet should be in plain English, without the article "The" at the beginning. I wasn't sure what you meant to suggest anything.

File: src/ansys_sphinx_theme/__init__.py [hunk 11]
```diff
@@ -285,6 +286,92 @@ def update_footer_theme(
...
+    context: Dict[str, Any],
+) -> None:
+    """Download the cheatsheet to the static directory.
+
+    Parameters
+    ----------
+    app : ~sphinx.application.Sphinx
+        Application instance for rendering the documentation.
+    cheatsheet_options : dict
+        Dictionary containing the cheatsheet options.
```
Comment (PipKat) [inline src/ansys_sphinx_theme/__init__.py:332]:
```suggestion
        Dictionary containing the cheat sheet options.
```

File: src/ansys_sphinx_theme/__init__.py [hunk 12]
```diff
@@ -285,6 +286,92 @@ def update_footer_theme(
...
+        Path containing the static folder.
+    context : dict
+        Dictionary containing the context for the page.
+    """
+    cheatsheet_url = cheatsheet_options.get("url", "")
+    cheatsheet_thumbnail = cheatsheet_options.get("thumbnail", "")
+    static_path = pathlib.Path(app.outdir) / static_folder[0]
+    context["cheatsheet_static_path"] = str(static_folder[0])
+
+    # Download cheatsheet file if URL is provided
```
Comment (PipKat) [inline src/ansys_sphinx_theme/__init__.py:343]:
```suggestion
    # Download cheat sheet file if URL is provided
```

File: src/ansys_sphinx_theme/__init__.py [hunk 13]
```diff
@@ -285,6 +286,92 @@ def update_footer_theme(
...
+    if cheatsheet_url:
+        download_file(cheatsheet_url, static_path)
+
+    # Download cheatsheet image if thumbnail URL is provided
```
Comment (PipKat) [inline src/ansys_sphinx_theme/__init__.py:347]:
```suggestion
    # Download cheat sheet image if thumbnail URL is provided
```

File: src/ansys_sphinx_theme/__init__.py [hunk 14]
```diff
@@ -285,6 +286,92 @@ def update_footer_theme(
...
+    if cheatsheet_thumbnail:
+        download_file(cheatsheet_thumbnail, static_path)
+
+
+def download_file(url: str, directory: pathlib.Path) -> None:
+    """
+    Download a file from the given URL and save it to the specified directory.
```
Comment (PipKat) [inline src/ansys_sphinx_theme/__init__.py:354]:
```suggestion
    Download a file from the given URL and save it to a given directory.
```

File: src/ansys_sphinx_theme/__init__.py [hunk 15]
```diff
@@ -285,6 +286,92 @@ def update_footer_theme(
...
+
+    Parameters
+    ----------
+    url : str
+        The URL of the file to download.
+    directory : pathlib.Path
+        The directory where the file will be saved.
```
Comment (PipKat) [inline src/ansys_sphinx_theme/__init__.py:361]:
```suggestion
        URL of the file to download.
    directory : pathlib.Path
        Directory to save the file to.
```

File: src/ansys_sphinx_theme/__init__.py [hunk 16]
```diff
@@ -285,6 +286,92 @@ def update_footer_theme(
...
+        file_path = directory / filename
+        with open(file_path, "wb") as file:
+            response = requests.get(url)
+            if response.status_code != 200:
+                raise FileNotFoundError(f"Failed to download file from {url}.")
+            file.write(response.content)
+
+
 def setup(app: Sphinx) -> Dict:
     """Connect to the sphinx theme app.
```
Comment (PipKat) [inline src/ansys_sphinx_theme/__init__.py:376]:
```suggestion
    """Connect to the Sphinx theme app.
```

File: src/ansys_sphinx_theme/theme/ansys_sphinx_theme/static/css/ansys_sphinx_theme.css [hunk 17]
```diff
@@ -1133,3 +1133,31 @@ a:visited,
 a:visited:hover {
   color: var(--pst-color-link-visted);
 }
+
+/* Cheatsheets sidebar */
```
Comment (PipKat) [inline src/ansys_sphinx_theme/theme/ansys_sphinx_theme/static/css/ansys_sphinx_theme.css:1137]:
```suggestion
/* Cheat sheet sidebar */
```

**Review by PipKat — APPROVED** [review]
@Revathyvenugopal162 Based on past Google searches and another one performed just now, "cheat sheet" is two words in the majority of cases. The titles of our existing cheat sheets also use it as two words. I propose that in the descriptive text (at least), we write it as two words.

File: src/ansys_sphinx_theme/__init__.py [hunk 18]
```diff
@@ -285,6 +286,80 @@ def update_footer_theme(
...
+    Parameters
+    ----------
+    app : ~sphinx.application.Sphinx
+        Application instance for rendering the documentation.
+    exception : Exception
+        Exception that was raised.
+    """
+    cheatsheet_url = cheatsheet_options.get("url", "")
+    cheatsheet_image = cheatsheet_options.get("image", "")
+    static_path = app.outdir / "_static"
```
[the remark this answers is not in the document]
Comment (RobPasMue) [reply src/ansys_sphinx_theme/__init__.py]: I like it much better this way - good one @Revathyvenugopal162!

**Review by RobPasMue — APPROVED** [review]
LGTM after all the team suggestions - awesome work @Revathyvenugopal162

**Comment by RobPasMue** [comment]
Let's try and get this merged and released before the Thursday meeting so that we can present it to the group!

**Comment by Revathyvenugopal162** [comment]
> Let's try and get this merged and released before the Thursday meeting so that we can present it to the group!

Thank you all for the review, Sure we can merge this and before the Thursday meeting. i will testing this branch in one/two library and merge the PR

**Comment by Revathyvenugopal162** [comment]
check the branch with pymechanical https://github.com/ansys/pymechanical/pull/638

**Comment by Revathyvenugopal162** [comment]
merging, thank you all.
